### ISSUE TITLE
Support for `Debug` trait implementation in `ClientWithMiddleware`

### ISSUE THREAD

**Issue by phosphene47** [TrueLayer/reqwest-middleware#20]
## Motivations

`ClientWithMiddleware` does not implement the `Debug` trait, which prevents other wrapper structs from automatically implementing the `Debug` trait. Consider this code for example:

```rust
#[derive(Debug)]
pub struct SomeApiClient {
    client: ClientWithMiddleware,
}
```

This code fails to compile because `SomeApiClient` cannot `derive(Debug)` when a field does not implement `Debug`.

## Solution

`ClientWithMiddleware` implements `Debug`.

## Alternatives

At the application level, write a `Debug` implementation for `SomeApiClient` by hand, ignoring the `client` field intentionally. This is quite tedious, repetitive and prone to mistakes.

## Additional context

Currently, `ClientWithMiddleware` is defined like this:

```rust
#[derive(Clone)]
pub struct ClientWithMiddleware {
    inner: reqwest::Client,
    middleware_stack: Box<[Arc<dyn Middleware>]>,
}
```

It's probably necessary to make `Debug` a supertrait of `Middleware`. If that is not possible, the `Debug` implementation of `ClientWithMiddleware` could instead skip the `middleware_stack` field, although I think hiding fields from the debug output should be avoided if possible.

**Comment by 06chaynes** [comment]
Opened pull request #21 to add these changes

**Comment by LukeMathWalker** [comment]
As I commented on in #21, making `Debug` a super trait of `Middleware` is a breaking change - we will not be merging it until we have another reason to release a new minor version.
If you want to implement `Debug` for `ClientWithMiddleware` now, I'd suggest opening a PR using the alternative strategy - skip the `middleware_stack` field.

**Comment by tl-rodrigo-gryzinski** [comment]
Fixed in #22
